### PR TITLE
[BUG]: Fixed docker compose up containers to be removed

### PR DESCRIPTION
**Pull request by Yaxhveer**
## Related Issue
  - Info about Issue or bug

Closes: #1533 

#### Describe the changes you've made
Docker trigger the start event when the application is again started.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

Please describe the tests(if any). Provide instructions how its affecting the coverage.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

|        Original         |          Updated           |
|:-----------------------:|:--------------------------:|
| **original screenshot** | <b>updated screenshot </b> |

### CODE REVIEW

**Comment by Yaxhveer** [comment]
@PranshuSrivastava Please review the PR

**Comment by PranshuSrivastava** [comment]
@Yaxhveer I think you have made the changes for the docker start command, that is listed in this issue #1539.  Were you able to replicate the error for this issue? You need to remove the need to remove containers after docker compose up fails to run once. For some reason when you start the containers with Keploy again, Keploy is not able to find the desired container.

**Review by PranshuSrivastava — COMMENTED** [review]
Please replicate the issue and try to fix it again.

**Comment by Yaxhveer** [comment]
@PranshuSrivastava Yeah I understand the issue correctly now, will update the changes by tomorrow.

**Comment by Yaxhveer** [comment]
@PranshuSrivastava Hey is detached mode for docker allowed in keploy command?

**Comment by Yaxhveer** [comment]
@PranshuSrivastava Kindly review the changes and let me know if it works.

File: pkg/clients/clients.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ type InternalDockerClient interface {
 	ConnectContainerToNetworks(containerName string, settings map[string]*network.EndpointSettings) error
 	ConnectContainerToNetworksByNames(containerName string, networkName []string) error
 	StopAndRemoveDockerContainer() error
+	StopAndRemoveDockerContainerByContainerID(containerID string) error
```
Comment (PranshuSrivastava) [inline pkg/clients/clients.go:14]: This name is too long please change it

File: pkg/hooks/launch.go [hunk 2]
```diff
@@ -283,6 +284,21 @@ func (h *Hook) processDockerEnv(appCmd, appContainer, appNetwork string, buildDe
 
 			err := h.runApp(appCmd, true)
 			h.logger.Debug("Application stopped with the error", zap.Error(err))
+
+			containerNames, composeErr := getContainerFromComposeFile("kdocker-compose.yaml")
```
Comment (PranshuSrivastava) [inline pkg/hooks/launch.go:288]: kdocker compose file is not created everytime, it is created only in the case when keploy-network is not mentioned in the docker compose file and keploy needs the keploy-network to run. So we create a new file called kdocker compose file where the network is keploy-network.

**Review by PranshuSrivastava — CHANGES_REQUESTED** [review]

File: pkg/hooks/launch.go (hunk 2 above)
Comment (Yaxhveer) [reply pkg/hooks/launch.go:288]: ok working on this

File: pkg/clients/clients.go (hunk 1 above)
Comment (Yaxhveer) [reply pkg/clients/clients.go:14]: ok

**Comment by Yaxhveer** [comment]
@PranshuSrivastava Please review the changes

File: pkg/clients/clients.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ type InternalDockerClient interface {
 	ConnectContainerToNetworks(containerName string, settings map[string]*network.EndpointSettings) error
 	ConnectContainerToNetworksByNames(containerName string, networkName []string) error
 	StopAndRemoveDockerContainer() error
+	StopAndRemoveDockerContainerByID(containerID string) error
```
Comment (PranshuSrivastava) [inline pkg/clients/clients.go:14]: This is still very long, can you change it to something like "stopandremovebyid" as we are obviously removing docker containers in this file

File: pkg/hooks/launch.go [hunk 4]
```diff
@@ -261,7 +262,7 @@ func (h *Hook) LaunchUserApplication(appCmd, appContainer, appNetwork string, De
 	}
 }
 
-func (h *Hook) processDockerEnv(appCmd, appContainer, appNetwork string, buildDelay time.Duration) error {
+func (h *Hook) processDockerEnv(appCmd, appContainer, appNetwork string, buildDelay time.Duration, isDockerComposeCmd bool) error {
```
Comment (PranshuSrivastava) [inline pkg/hooks/launch.go:265]: why do we need this boolean, can't we just check if the containers are running and then remove them if needed.

**Review by PranshuSrivastava — CHANGES_REQUESTED** [review]

File: pkg/clients/clients.go (hunk 3 above)
Comment (Yaxhveer) [reply pkg/clients/clients.go:14]: ok

**Comment by charankamarapu** [comment]
Hey @Yaxhveer we have refactored the main code base. But the docker related testing is not yet done. We will notify you once that is done and then we continue with the PR if the issue is not solved by the refactored changes.

**Comment by Yaxhveer** [comment]
@PranshuSrivastava The PR has been updated according to the current codebase. Kindly the review the PR and let me know if it require any changes.
Still there is a issue regarding blocking and panic in record mode as I specified [here](https://github.com/keploy/keploy/issues/1795).

**Comment by gouravkrosx** [comment]
@Yaxhveer Please make a fresh PR for the same issue and close this one.

**Pull request closed without merging**